daml-script/daml3/Daml/Script/Questions/Submit.daml

```diff
@@ -10,168 +11,258 @@ module Daml.Script.Questions.Submit
+
+-- | Allows for concurrent submission of transactions, using an applicative, similar to Commands.
+-- Concurrently takes a computation in `ConcurrentSubmits`, which supports all the existing `submit` functions
+-- that `Script` supports. It however does not implement `Action`, and thus does not support true binding and computation interdependence
+concurrently : HasCallStack => ConcurrentSubmits a -> Script a
+concurrently submissions = fmap submissions.continue $ lift $ Submit with
+  submissions = submissions.submits
+
+instance ScriptSubmit Script where
+  submitInternal : HasCallStack => SubmitOptions -> ErrorBehaviour -> Commands a -> Script (Either (Optional SubmitError) (a, Optional TransactionTree))
+  submitInternal opts errorBehaviour cmds = concurrently $ submitInternal opts errorBehaviour cmds
+
+-- assertFromRight for mustsucceed. Error path should never be called, as scala handles it.
+mustSucceed : Either a b -> b
+mustSucceed = \case
+  Left _ -> error "MustSucceed Failed"
+  Right r -> r
+
+-- assertFromLeft for mustfail. Error path should never be called, as scala handles it.
+mustFail : Either a b -> a
+mustFail = \case
+  Left e -> e
+  Right _ -> error "MustFail Succeeded"
 
--- There are a few different options for passing the parties here:
+-- ##### Main API #####
+-- | Equivalent to `submit` but returns the result and the full transaction tree.
+submitResultAndTree : (HasCallStack, ScriptSubmit m, IsSubmitOptions o) => o -> Commands a -> m (a, Optional TransactionTree)
+submitResultAndTree opts cmds = mustSucceed <$> submitInternal (toSubmitOptions opts) MustSucceed cmds
+
+-- | Equivalent to `trySubmit` but returns the result and the full transaction tree.
+trySubmitResultAndTree : (HasCallStack, ScriptSubmit m, IsSubmitOptions o) => o -> Commands a -> m (Either SubmitError (a, Optional TransactionTree))
+trySubmitResultAndTree opts cmds = first getError <$> submitInternal (toSubmitOptions opts) Try cmds
+
+-- | HIDE
+submitWithError_ : (HasCallStack, ScriptSubmit m, IsSubmitOptions o) => o -> Commands a -> m (Optional SubmitError)
+submitWithError_ opts cmds = mustFail <$> submitInternal (toSubmitOptions opts) MustFail cmds
+
+-- | Equivalent to `submitMustFail` but returns the error thrown.
+submitWithError : (HasCallStack, ScriptSubmit m, IsSubmitOptions o) => o -> Commands a -> m SubmitError
+submitWithError opts cmds = getError <$> submitWithError_ opts cmds
+
+getTransactionTree : Optional TransactionTree -> TransactionTree
+getTransactionTree = \case
+  None -> error "No transaction tree returned! The JSON Ledger Client does not support submitTree."
+  Some t -> t
+
+getError : Optional SubmitError -> SubmitError
+getError = \case
+  None -> error "No transaction tree returned! The JSON Ledger Client does not support submit for error."
+  Some e -> e
+
+-- | `submit p cmds` submits the commands `cmds` as a single transaction
+-- from party `p` and returns the value returned by `cmds`.
```

Comment:
This is now out of date, should add a fairly big explanation here about how to use `SubmitOptions`

Reply:
I don't think we should spend too much time on documenting this until things are stable, but could you please add a TODO so we don't forget to do it?

Reply:
Ah right, I'll make a ticket

Reply:
[#18035](https://github.com/digital-asset/daml/issues/18035)